### PR TITLE
Add maximum expected displacement (Emax) as a path metric

### LINKED ISSUE #905: [Metrics] Add function for computing E_max (maximum expected displacement)
**Is your feature request related to a problem? Please describe.**
$E_{\max}$ is a scale-independent straightness measure based on eigenvalue decomposition of the mean step rotation matrix. Unlike the straightness index ($D/L$), which is sensitive to trajectory length, $E_{\max}$ characterises the inherent directional persistence of the movement process itself. Computes how far an animal is expected to travel given its typical turning behavior and step lengths. It is one of the core trajectory complexity metrics identified in the [trajr paper](https://doi.org/10.1111/eth.12739)

### Definition
 
Given the mean cosine $\bar{c}$ and mean sine $\bar{s}$ of turning angles, define the mean step rotation matrix:
 
$$\mathbf{M} = \begin{pmatrix} \bar{c} & -\bar{s} \\ \bar{s} & \bar{c} \end{pmatrix}$$
 
Then:
 
$$E_{\max}^{(a)} = \lambda_{\max}\left[(\mathbf{I} - \mathbf{M})^{-1}\right]$$
 
where $\lambda_{\max}$ denotes the largest eigenvalue.
 
- $E_{\max}^{(a)}$ is dimensionless (scale-independent).
- $E_{\max}^{(b)} = E_{\max}^{(a)} \cdot \bar{p}$ gives the maximum expected displacement in spatial units (where $\bar{p}$ is the mean step length).
 
**Range:** $[0, \infty)$. Values near 0 indicate highly sinuous paths; larger values indicate straighter paths.
 
### Alternative names
 
Also referred to as: maximum expected displacement, Cheung's Emax, Emax-a (dimensionless) / Emax-b (spatial).

### Pseudocode
 
```
function compute_emax(data, emax_b=False, nan_policy="ffill"):
    # 1. Compute turning angles (depends on movement#833)
    theta = compute_turning_angle(data)
 
    # 2. Compute step lengths
    step_lengths = compute_norm(compute_forward_displacement(data))
 
    # 3. Handle NaN per nan_policy
 
    # 4. Compute mean cosine and sine of turning angles
    c_bar = nanmean(cos(theta))
    s_bar = nanmean(sin(theta))
 
    # 5. Build mean rotation matrix M
    M = [[c_bar, -s_bar],
         [s_bar,  c_bar]]
 
    # 6. Compute (I - M)^{-1}
    I_minus_M = eye(2) - M
    inv_matrix = inv(I_minus_M)
 
    # 7. Emax-a = largest eigenvalue of inv_matrix
    eigenvalues = eigvals(inv_matrix)
    emax_a = max(real(eigenvalues))
 
    # 8. Optionally scale by mean step length
    if emax_b:
        return emax_a * nanmean(step_lengths)
    return emax_a
```

**Parameters:**
- `emax_b`: If `True`, return $E_{\max}^{(b)}$ (spatial units). Default `False` returns $E_{\max}^{(a)}$ (dimensionless).
 
**Returns:** A scalar per individual/keypoint, with `time` and `space` dimensions removed.

### References
 
- **Primary:** Cheung, A., Zhang, S., Stricker, C. & Srinivasan, M. V. (2007). Animal navigation: the difficulty of moving in a straight line. *Biological Cybernetics* 97(1), 47-61. https://doi.org/10.1007/s00422-007-0158-0
- **trajr implementation:** `TrajEmax` — McLean & Skowron Volponi (2018). https://doi.org/10.1111/eth.12739
 
### Context
 
Opened following discussion with @niksirbi on [Zulip](https://neuroinformatics.zulipchat.com/#narrow/channel/406001-Movement/topic/Trajectory.20Complexity.20Metrics.20.28Issue.20.23406.2F.23835.29). Related to https://github.com/neuroinformatics-unit/movement/issues/406. Depends upon #833

### PR DESCRIPTION
**Pull request by vybhav72954**
Adds `compute_maximum_expected_displacement` (Cheung 2007's E_max) to `movement.kinematics.path`.

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
E_max is a straightness measure that quantifies how far an animal is expected to travel given its observed distribution of turning angles and step lengths; assuming it navigates without an external directional reference (e.g. no compass, no landmark)

**What does this PR do?**

  - `E_max⁽ᵃ⁾ = c̄ / (1 − c̄)` with `c̄ = nanmean(cos θ)` over turning angles (dimensionless), and `E_max⁽ᵇ⁾ = p̄ ·
  E_max⁽ᵃ⁾` scaled by mean step length (spatial units).
  - `in_spatial_units=True` (default) → spatial variant; `False` → dimensionless. Default follows your steer in #905
  (most users stay within one species/setup).
  - Light-touch per the 2026-06-12 community call: no `min_step_length`, `nan_warn_threshold`, or `nan_policy`.
  `in_degrees` is deliberately **not** exposed — `cos θ` requires radians.

## References
- Closes #905
- Related to #406 
- Link to [Zulip](https://neuroinformatics.zulipchat.com/#narrow/channel/406001-Movement/topic/Potential.20trajectory.20complexity.20metrics/with/596429211)

Cheung, A., Zhang, S., Stricker, C. & Srinivasan, M. V. (2007). Animal navigation: the difficulty of moving in a straight line. *Biological Cybernetics* 97(1), 47-61. https://doi.org/10.1007/s00422-007-0158-0

## How has this PR been tested?
Created new parametrized test functions in `tests/test_unit/test_kinematics/test_path.py`:

## Is this a breaking change?
No

## Does this PR require an update to the documentation?

Added a self contained docstring

### Naming — question + my suggestion
Following #1021 (`directional_change` → `turning_angle_rate`, moving off opaque paper labels), I've named the function
**`compute_maximum_expected_displacement`** rather than `compute_path_emax`. It's self-documenting, matches
`long_name="Maximum Expected Displacement"`, and sits in the same vocabulary as `compute_*_displacement`; `E_max` / 
trajr's `TrajEmax` stay in the docstring first line + Notes for discoverability. The scale toggle is **`in_spatial_units`** (mirrors `in_degrees`), replacing the `emax_b` name you flagged.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: movement/kinematics/path.py [hunk 1]
```diff
@@ -393,6 +393,143 @@ def compute_directional_change(
     return dc
 
 
+def compute_maximum_expected_displacement(
+    data: xr.DataArray,
+    in_spatial_units: bool = True,
```
Comment (niksirbi) [inline movement/kinematics/path.py:531]: I like your choice of `in_spatial_units` as the kwarg name, consistent with `in_degrees`!

File: movement/kinematics/path.py [hunk 2]
```diff
@@ -393,6 +393,143 @@ def compute_directional_change(
...
+       :math:`\theta` is the signed rotation from the backward
+       displacement step arriving at ``t-1`` to the step arriving at
+       ``t``, computed via
+       :func:`movement.utils.vector.compute_signed_angle_2d`. The first
+       two time steps have no defined turning angle and so do not
+       contribute.
+    2. **Radians only.** The turning angles must be in radians for
+       :math:`\cos\theta` to be meaningful, so the ``in_degrees`` flag of
+       :func:`compute_turning_angle` is deliberately *not* exposed here;
+       passing angles in degrees would silently corrupt the result.
```
Comment (niksirbi) [inline movement/kinematics/path.py]: Not a stongly held opinion of mine, but I'd lean towards removing this 2nd note. It's good to know, but not particularly actionable for the user, since they are never presented with that option, and the result of this function is anyway not an angle (so saying 'radians only' could be slightly confusing).

File: movement/kinematics/path.py [hunk 3]
```diff
@@ -393,6 +393,143 @@ def compute_directional_change(
...
+    3. **Range.** :math:`E_{\max}^{(a)} \in [-0.5, \infty)`. Negative
+       values arise for highly sinuous paths, where the mean cosine
+       :math:`\bar{c}` is itself negative.
```
Comment (niksirbi) [inline movement/kinematics/path.py]: What you are saying here is not wrong, but I would phrase it a bit more specifically:

```suggestion
    3. **Range.** :math:`E_{\max}^{(a)} \in [-0.5, \infty)`. While
       highly sinuous paths actually have values approaching 0, 
       negative values specifically arise for trajectories with a
       systematic backward-turning bias where the mean cosine
       :math:`\bar{c}` is itself negative.
```

**Review by niksirbi — APPROVED** [review]
This is an excellent PR @vybhav72954!

The implementation is mathematically correct and you've also done a very thorough job with the docstring and the tests. You've adhered to all of our conventions and even thought of non-obvious edge cases that hadn't occurred to me (negative Emax values).

The CI failures are likely fixable by rebasing this PR onto the `main` branch.

I'm approving this PR with a few optional comments, but I'm holding off the merging for now because I'd like people's opinions on this function's name.

While `compute_maximum_expected_displacement` is correct and matches the literature, it has two drawbacks:

1. It's not obvious from the function name that this is a measure of straightness. One wouldn't look for it unless one already knew of that metric. I think we can alleviate that drawback by cross-referencing the new functions from the docstring of `compute_path_straightness` (in that function's 'See Also' section, or perhaps also adding a note about the advantages of Emax over D/L). We could also mention it in `compute_directional_change`'s 'See also' section.
2. It's not obvious that it's a 'path' metric. Currently all motion quantification functions are importable `from movement.kinematics`, and appear listed together in the [API reference](https://movement.neuroinformatics.dev/latest/api/movement.kinematics.html). With this status quo, the mention of 'path' in the function name is the only indicator that a metric has anything to do with trajectories/paths. We could call it `compute_path_emax`, which would be more consistent with other metrics in the same category, but I imagine that `emax` is similarly opaque to most users. `compute_path_max_expected_displacement` is somewhat long of a name.
   
I am quite possibly overthinking this, but I'd like to hear other opinions from @neuroinformatics-unit/movement-active-devs on the name. 
Perhpas mentioning the new function in the 'See Also' sections of related metrics is enough to aid discoverability for now, without renaming it. 

Future possible alleviating measures would be to:
- expand the module-level docstring of `movement.kinematics` to provide some logical grouping of all the relevant functions, or split the `kinematics` module into several modules
- Write an example walking through multiple path-level metrics (planned anyway).

**Comment by vybhav72954** [comment]
@niksirbi Thanks for the inputs, I concur with the naming convention, it's best to wait for others' suggestions, meanwhile I have made small updates to the `path.py` docstrings.

Let me know if i can make any furtehr changes.

**Review by niksirbi — COMMENTED** [review]
Hey @vybhav72954, sorry for the delay on our side. I'm hoping to clarify the question about the function's name tomorrow and get back to you in the next days.

Apart from that, I've also reviewed and merged #981 today. Now we need to rebase this PR on the main branch, and I'm afraid you'll have to deal with some merge conflicts, and change a few things to be consistent with #981.

Here's a list of things that will need updating:

- Change the PR's name to something like "Add maximul expected displacement as a path metric". No need to mention the issue number in the PR title, as it's already present in the PR description (and auto-linked by GitHub).
- `_validate_time_points` now allows you to specify  `min_points=3`, and the Emax function should call the validator with that kwarg (>=3 points are needed for a single valid turning angle). The corresponding test will need to be updated accordingly.
- While reviewing #981 I realised that xarray suppresses zero-division runtime warnings be default, so the guard against zero division will not be necessary. Adapt this bit to match what `compute_path_sinuosity` does
- Use `xr.apply_ufunc(np.cos, theta)` instead of `cast("xr.DataArray", np.cos(theta))`, again following the example of `compute_path_sinuosity`
- Add the `nan_warn_threshold` kwarg (but still no NaN policy and no `min_step_length`). Some of our path metrics have `nan_warn_threshold` and some don't, but there is an implicit logic to it. Path length, straightness and sinuosity are all metrics that summarise the entire trajectory; many missing values could lead to a mis-estimate. On the other hand, directional change and path deviation return 1 value per step, and there we can just let nans propagate naturally per step. By that logic, Emax should include `nan_warn_threshold` because it's a summary metric for the whole trajectory. Let me know if you disagree.
- The first docstring note refers to `compute_turning_angle` (which is good), but then also proceeds to explain the utilities underlying that function. Explaining the internals of internals may be going too far. Simplify the first note by removing the sentence about `compute_turning_angle` internals.
- Docstring section order here is Notes → References → See Also → Examples, while `compute_path_sinuosity` follows See Also → Notes → References → Examples (numpydoc canonical). Change Emax docstring order to numpydoc canonical as well. I'm aware we are not consistent with this throughout the codebase, but I plan to open a separate general issue for that. (now opened at #1048).
- Cross-reference the two functions (sinuosity and Emax) in each others' See Also sections. Perhaps also reference sinuosity from path straightness (for completeness).

One thing I'd explicitly *not* do: factor out the shared `mean(cos(turning angle))` / `mean(step length)` into a helper. There are only two call sites with two lines each. The helper costs more than it saves for now. If a more callers (path metrics) are added, that's the moment to do it.

Do let me know if any of that doesn't make sense, or if you find any more discrepancies.

**Comment by vybhav72954** [comment]
Thanks for the inputs niko. I concur with the `nan` policy, it makes total sense, i had something similar in mind!

On the docstring order, I noticed you have initiated a new issue and am assuming the team is working hard to remove all the redundancies across the Repo. I will clean this one up.

The rest of the updates are mostly mechanical in nature, I believe I will get to them, in case I have any questions. I will get back!

**Comment by niksirbi** [comment]
Hey @vybhav72954, we discussed this function's name with @lochhh in today's community call.

Our shared consensus was that `compute_path_emax` is the least bad option for now (it's consistent with existing conventions), so let's rename it n this PR.

Some of the discoverability problems we were worried about will go away in future when we:

- Write an [example](https://movement.neuroinformatics.dev/latest/examples/index.html) showcasing existing path metrics, and/or
- Separate `movement.path` from `movement.kinematics` (we haven't yet made a decision on that)

Apart from that, could you also document the current convention about `nan_warn_threshold` (that I outlined above) in the module-level docstring of `path.py`? This should help developers of future `path.py` functions make a decision on whether to expose `nan_warn_threshold` in any a given new function (**Yes** if the metric returns a scalar per path, **No** if it returns a time-varying quantity).

Those were the only two novel relevant action items.

**Comment by vybhav72954** [comment]
@niksirbi 
Thanks for relaying the call's decision happy to go with `compute_path_emax`

Done in this PR: renamed the function (definition, the `__init__.py` export + `__all__`, the cross-references in the other path metrics' See Also sections, and the tests, including the test names)

I kept the literature term front-and-centre in the docstring summary and Notes/References, so it stays discoverable

I've also documented the `nan_warn_threshold` convention in the module-level docstring of `path.py`

I agree on the `nan_warn_threshold`. Since `E_max` collapses the whole trajectory into a single scalar, a high proportion of missing values can silently skew it, so the warning is warranted here.  I've added the `nan_warn_threshold` parameter (default 0.2) and the `_warn_about_nan_proportion` call

**Review by niksirbi — APPROVED** [review]
Thanks for thoroughly dealing with all my comments @vybhav72954, I have no additional requests. 

This PR is ready to merge now! 🎉
